pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -3411,18 +3415,38 @@ private long readLongValue(byte[] bytes, int oid, long minVal, long maxVal, Stri
         val = ByteConverter.int8(bytes, 0);
         break;
       case Oid.FLOAT4:
-        val = (long) ByteConverter.float4(bytes, 0);
+        float f = ByteConverter.float4(bytes, 0);
+        // for float values we know to be within values of long, just cast directly to long
+        if (f <= LONG_MAX_FLOAT && f >= LONG_MIN_FLOAT) {
+          val = (long) f;
+        } else {
+          throw new PSQLException(GT.tr("Bad value for type {0} : {1}", targetType, f),
+              PSQLState.NUMERIC_VALUE_OUT_OF_RANGE);
+        }
         break;
       case Oid.FLOAT8:
-        val = (long) ByteConverter.float8(bytes, 0);
+        double d = ByteConverter.float8(bytes, 0);
+        // for double values within the values of a long, just directly cast to long
+        if (d <= LONG_MAX_FLOAT && d >= LONG_MIN_FLOAT) {
+          val = (long) d;
+        } else {
+          throw new PSQLException(GT.tr("Bad value for type {0} : {1}", targetType, d),
+              PSQLState.NUMERIC_VALUE_OUT_OF_RANGE);
+        }
         break;
       case Oid.NUMERIC:
         Number num = ByteConverter.numeric(bytes);
-        if (num instanceof  BigDecimal) {
-          val = ((BigDecimal) num).setScale(0 , RoundingMode.DOWN).longValueExact();
+        if (num instanceof BigDecimal) {
+          if ( ((BigDecimal) num).compareTo(LONG_MAX_BD) <= 0 && ((BigDecimal) num).compareTo(LONG_MIN_BD) >= 0 ) {
```

Comment:
This differs from string based parsing which first [converts to a BigInteger](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html#toBigInteger--). This conversion truncates any fractional value.
So a numeric which is effective Long.MAX + 0.9 should return Long.MAX, but here would fail.

Reply:
What do you propose here then ?

Reply:
Changing the code to
```
    if (num instanceof BigDecimal) {
          val = ((BigDecimal) num).setScale(0 , RoundingMode.DOWN).longValueExact();
          if ( val <= Long.MAX_VALUE && val >= Long.MIN_VALUE ) {
            return val;
``` 
Throws an exception `at java.math.BigDecimal$LongOverflow.check(BigDecimal.java:3152)`

Reply:
https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L3223-L3232

```
          BigInteger i = ((BigDecimal) num).toBigInteger();
          int gt = i.compareTo(LONGMAX);
          int lt = i.compareTo(LONGMIN);

          if (gt > 0 || lt < 0) {
            throw new PSQLException(GT.tr("Bad value for type {0} : {1}", "long", s),
                PSQLState.NUMERIC_VALUE_OUT_OF_RANGE);
          }
          val = i.longValue();
```